lib/python/pyflyby/_saveframe.py

```diff
@@ -0,0 +1,1034 @@
+
+    :param exception_obj:
+      The exception raised by the user's code.
+    :return:
+      An `ExceptionInfo` object.
+    """
+    try:
+        tb = (
+            traceback.format_exception(
+                type(exception_obj), exception_obj, exception_obj.__traceback__))
+    except Exception as err:
+        _SAVEFRAME_LOGGER.warning(
+            f"Error while formatting the traceback. Error: {err!a}")
+        tb = "Traceback couldn't be formatted"
+    exception_info = ExceptionInfo(
+        exception_string=str(exception_obj),
+        exception_full_string=f'{exception_obj.__class__.__name__}: {exception_obj}',
+        exception_class_name=exception_obj.__class__.__name__,
+        exception_class_qualname=exception_obj.__class__.__qualname__,
+        exception_object=exception_obj,
+        traceback=tb
+    )
+    return exception_info
+
+
+def _get_qualname(frame):
+    """
+    Get fully qualified name of the function for the ``frame``.
+
+    In python 3.10, ``co_qualname`` attribute is not present, so use ``co_name``.
+    """
+    return (frame.f_code.co_qualname if hasattr(frame.f_code, "co_qualname")
+            else frame.f_code.co_name)
+
+def _get_frame_repr(frame):
+    """
+    Construct repr for the ``frame``. This is used in the info messages.
+
+    :param frame:
+      The frame object.
+    :return:
+      The string f'File: {filename}, Line: {lineno}, Function: {function_qualname}'
+    """
+    return (f"'File: {frame.f_code.co_filename}, Line: {frame.f_lineno}, "
+            f"Function: {_get_qualname(frame)}'")
+
+
+def _get_frame_local_variables_data(frame, variables, exclude_variables):
+    """
+    Get the local variables data of the ``frame``.
+
+    :param frame:
+      The frame object
+    :param variables:
+      Local variables to be included.
+    :param exclude_variables:
+      Local variables to be excluded.
+    :return:
+      A dict containing the local variables data, with the key as the variable
+      name and the value as the pickled local variable value.
+    """
+    # A dict to store the local variables to be saved.
+    local_variables_to_save = {}
+    all_local_variables = frame.f_locals
+    for variable in all_local_variables:
+        # Discard the variables that starts with '__' like '__eq__', etc., to
+        # keep the data clean.
+        if variable.startswith('__'):
+            continue
+        if variables and variable not in variables:
+            continue
+        if exclude_variables and variable in exclude_variables:
+            continue
+        try:
+            pickled_value = pickle.dumps(
+                all_local_variables[variable], protocol=PICKLE_PROTOCOL)
+        except Exception as err:
+            _SAVEFRAME_LOGGER.warning(
+                f"Cannot pickle variable: {variable!a} for frame: "
+                f"{_get_frame_repr(frame)}. Error: {err!a}. Skipping this "
+                f"variable and continuing.")
```

Comment:
In general you never want to use f-strings with loggers for multiple reasons:
 - the formatting when using the `.meth("template %s", argument)` syntax can be lazy if the argument object __repr__ or similar is expensive and the logger level low enoughit has less perf impact. 
 - logger handler can be set to store structured logging,  in which case the template and argument are stored in say different columns, and you can do query like "show me all the messages with the same template message" or do some other fancy stuff. 

Here I don't think it matters, but it can be better to show good practices. 

You can find some linters with rules preventing f-strings in loggers:

https://docs.astral.sh/ruff/rules/logging-f-string/

for example.